### PR TITLE
Major update

### PR DESCRIPTION
**Pull request by clane9**
- Add utils for creating datasets and data loaders supporting in memory data, timm data augmentation, cuda prefetching.
- Add utils for computing metrics and generating diagnostic figures.
- Rewrite model implementation to support architecture ablations.
- Add train script based on the timm train script supporting AMP, `torch.compile`, gradient accumulation, wandb.
- Update training playground notebook.

### CODE REVIEW

**Comment by clane9** [comment]
@IhabBendidi if you could take a quick look and see if you have any suggestions, it would be very helpful!

**Review by IhabBendidi — APPROVED** [review]
Hey @clane9 , the code is very good, I have looked through it and its very complete. Its good to go. 

I have a question however as I don't use the stability cluster but another cluster : Does it have job time limits and automatic requeuing of jobs? if so, do we need a function for that?

**Comment by clane9** [comment]
Thanks @IhabBendidi! I'm also using a different cluster for now. I think if we can it would be good to keep the train script agnostic to what cluster is used. We can revisit this later if integrating some job scheduling functionality would be useful. I've previously used [submitit](https://github.com/facebookincubator/submitit) for that sort of thing.
